File: 1290. Convert Binary Number in a Linked List to Integer/convert_binary_number_in_a_linked_list_to_integer.cpp
// T.C.->O(n)
// S.C.->O(1)
// For both Approaches

// Bit Manipulation
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int getDecimalValue(ListNode* head) {
        int result = 0;
        while (head != nullptr) {
            // Left shift the result by 1 bit to make space for the next bit, and then perform a bitwise OR with the current node's value to add the bit to the number
            result = (result << 1) | head->val;
            head = head->next;
        }
        return result;
    }
};


// Math
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int getDecimalValue(ListNode* head) {
        int ans = 0;
        for (; head; head = head->next)
            ans = ans * 2 + head->val;
        return ans;
    }
};

File: 1290. Convert Binary Number in a Linked List to Integer/convert_binary_number_in_a_linked_list_to_integer.py
# T.C.->O(n)
# S.C.->O(1)
# For both Approaches

# Bit Manipulation
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def getDecimalValue(self, head: Optional[ListNode]) -> int:
        result = 0
        while head:
            # Bitwise left shift result by 1 (equivalent to multiplying by 2) and perform bitwise OR with the current node's value to append it to the binary number we are building.
            result = (result << 1) | head.val
            head = head.next
        return result


# Math
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def getDecimalValue(self, head: Optional[ListNode]) -> int:
        ans = 0
        while head:
            ans = ans * 2 + head.val
            head = head.next
        return ans
